packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart

```diff
@@ -0,0 +1,753 @@
+       assert(reverseCrossAxis != null);
+
+  /// The number of children in the cross axis.
+  final int crossAxisCount;
+
+  /// The width in logical pixels of the day child widgets.
+  final double dayChildWidth;
+
+  /// The width in logical pixels of the edge child widgets.
+  final double edgeChildWidth;
+
+  /// Whether the children should be placed in the opposite order of increasing
+  /// coordinates in the cross axis.
+  ///
+  /// For example, if the cross axis is horizontal, the children are placed from
+  /// left to right when [reverseCrossAxis] is false and from right to left when
+  /// [reverseCrossAxis] is true.
+  ///
+  /// Typically set to the return value of [axisDirectionIsReversed] applied to
+  /// the [SliverConstraints.crossAxisDirection].
+  final bool reverseCrossAxis;
+
+  /// The number of logical pixels from the leading edge of one row to the
+  /// leading edge of the next row.
+  double get _rowHeight {
+    return _monthItemRowHeight + _monthItemSpaceBetweenRows;
+  }
+
+  /// The height in logical pixels of the children widgets.
+  double get _childHeight {
+    return _monthItemRowHeight;
+  }
+
+  @override
+  int getMinChildIndexForScrollOffset(double scrollOffset) {
+    return crossAxisCount * (scrollOffset ~/ _rowHeight);
+  }
+
+  @override
+  int getMaxChildIndexForScrollOffset(double scrollOffset) {
+    final int mainAxisCount = (scrollOffset / _rowHeight).ceil();
+    return math.max(0, crossAxisCount * mainAxisCount - 1);
+  }
+
+  double _getCrossAxisOffset(double crossAxisStart, bool isPadding) {
+    if (reverseCrossAxis) {
+      return
+        ((crossAxisCount - 2) * dayChildWidth + 2 * edgeChildWidth) -
+        crossAxisStart -
+        (isPadding ? edgeChildWidth : dayChildWidth);
+    }
+    return crossAxisStart;
+  }
+
+  @override
+  SliverGridGeometry getGeometryForChildIndex(int index) {
+    final int adjustedIndex = index % crossAxisCount;
+    final bool isEdge = adjustedIndex == 0 || adjustedIndex == crossAxisCount - 1;
+    final double crossAxisStart = math.max(0, (adjustedIndex - 1) * dayChildWidth + edgeChildWidth);
+
+    return SliverGridGeometry(
+      scrollOffset: (index ~/ crossAxisCount) * _rowHeight,
+      crossAxisOffset: _getCrossAxisOffset(crossAxisStart, isEdge),
+      mainAxisExtent: _childHeight,
+      crossAxisExtent: isEdge ? edgeChildWidth : dayChildWidth,
+    );
+  }
+
+  @override
+  double computeMaxScrollOffset(int childCount) {
+    assert(childCount >= 0);
+    final int mainAxisCount = ((childCount - 1) ~/ crossAxisCount) + 1;
+    final double mainAxisSpacing = _rowHeight - _childHeight;
+    return _rowHeight * mainAxisCount - mainAxisSpacing;
+  }
+}
+
+/// Displays the days of a given month and allows choosing a date range.
+///
+/// The days are arranged in a rectangular grid with one column for each day of
+/// the week.
+class _MonthItem extends StatelessWidget {
+  /// Creates a month item.
+  _MonthItem({
```

Comment:
const?

Reply:
Unfortunately the asserts use DateTime methods which prevent this. I could move them to the body of the constructor, but as this was an internal API, it didn't seem worth it.

Reply:
The assert provide more safety than the const provides readability. Sounds OK.